### ISSUE TITLE
Infinite looping of VCI page

### ISSUE THREAD

**Issue by mrmin123** [ClinGen/clincoded#731]
Cannot reproduce on localhost, but reproducible on instances.

1. Create variant via VCI interface
2. Reload the page on VCI page to begin infinite loop

Console log:
![image](https://cloud.githubusercontent.com/assets/4326866/16322014/2d18abac-3955-11e6-8dfd-def322e2a0f1.png)

I believe it's the Invariant Violation is the cause of the loop. I'm hoping that #730 will fix the issue but this is not guaranteed, which is the reason for this ticket.

**Comment by jimmyzhen** [comment]
In the process of debugging the infinite loop problem upon reloading the page multiple times, I was able to reproduce the issue on the latest `dev` version instance on AWS (e.g. https://dev-061716.demo.clinicalgenome.org).

By using the ClinVar ID `139214` and navigating to the "Evidence Hub" interface, I noticed that the Basic Information tab no longer displays data on initial page load like I expected from my prior commits (this behavior can be consistently reproduced in my local instance):
![image](https://cloud.githubusercontent.com/assets/6956310/16437149/13c845a6-3d59-11e6-8709-3f1c205b2a92.png)

Then reloading the page would seemingly invoke data fetching and result in the expected display of data on the Basic Information tab.

Following the prior page reload completes, another page reload would incur a seemingly "infinite looping" of page loads. I used the word *seemingly* because the looping eventually stopped when it reached the console log messages pertinent to *non-bundled* files:
![image](https://cloud.githubusercontent.com/assets/6956310/16437476/36485cc6-3d5c-11e6-9673-1afa7e928a85.png)

Even my own debugging instance demonstrated a similar behavior, in which the looping eventually stopped when the console log messages were no longer pertinent to bundled files:
![image](https://cloud.githubusercontent.com/assets/6956310/16437524/96d9d736-3d5c-11e6-8ee8-ebcc575657f7.png)

The `Invariant Violation` error appears to be present in both of the non-looping and looping page reloads.

Because the looping page reload is pertinent to an AWS instance and not present in my local instance, the debugging process has been time-consuming and so far hasn't been able to identify the culprit. I will continue and try other methods.

**Comment by jimmyzhen** [comment]
In regards to the problem in which data is not displayed on **Basic Info** tab upon initial page load incurred by navigating from variant selection page, I've found that the culprit appears to be the order of loading components in the app. In the case of VCI, the child component (*Basic Info*) was mounted prior to the parent component that relies on its own lifecycle method to change the state and pass the state value onto its child component as a props.

Due to this component loading order, the child component received the initial `false` state as a props (from the parent) prior to the parent component changing the state to `true` for the props. As such, the `this.state.shouldFetchData === true` condition in the `componentWillReceiveProps` lifecycle method used by the **Basic Info** tab fails to invoke the data fetching methods.

I have updated the code with a simple fix to address this problem and pushed my commits to the repo (e.g. https://github.com/ClinGen/clincoded/compare/731_jz_infinite_loop).

**Comment by mrmin123** [comment]
Hey @jimmyzhen, thanks for tracking this down. If you could, can you also remove the reliance on `shouldFetchData` from the other tabs as well, if it's superfluous/buggy?

I'd also like to suggest taking note of [this](https://facebook.github.io/react/blog/2016/01/08/A-implies-B-does-not-imply-B-implies-A.html) when using `componentWillReceiveProps`. Most relevant:
>Some people assume this means "if componentWillReceiveProps is called, then the props must have changed", but that conclusion is logically incorrect.

You should implement a check to make sure that the received `prop.data` is not the same as the existing `prop.data`, otherwise the `get...Data()` methods will be called multiple times.

Also, just a heads up that you might have some merge conflicts with [PR 735](https://github.com/ClinGen/clincoded/compare/dev...732_mc_population_tab?expand=1) as I added a `componentDidMount` method and made slight changes to the `componentWillReceiveProps` methods. Feel free to change these as necessary.

**Comment by jimmyzhen** [comment]
@mrmin123, I will remove the dependency on `shouldFetchData` from other tabs as you suggested, as we can now shift the dependency to `props.data`.

Also, since Karen already merged your PR, I will merge it into my branch and base my changes from there.

In regards to invoking the data fetching methods only when `nextProps.data !== this.props.data` in the context of the `componentWillReceiveProps` lifecycle method, I found that both `nextProps.data` and `this.props.data` have the same value upon the completion of initial rendering in the client. This behavior appears to be consistent when I tried selecting a different ClinVar ID or changing the UUID in the URL directly. As such, I think I will use a boolean state to invoke data fetching if `this.props.data` returns value. Then I will reset the state in the `componentWillUnmount` method.
